### PR TITLE
Fix Kafka output panic at startup

### PR DESCRIPTION
**Pull request by cmacknz**
- Closes https://github.com/elastic/beats/issues/41823

In https://github.com/elastic/beats/pull/40794 the Connectable interface was changed to include a `context.Context` argument, causing all clients with a Connect() that weren't updated to have a `Context(context.Context)` method instead to fail the NetworkClient interface check in https://github.com/elastic/beats/blob/367d94e1dc40a82713d651735eac7921bb584474/libbeat/publisher/pipeline/client_worker.go#L64-L70

This made it so that the Kafka and Redis outputs no longer counted as NetworkClients, bypassing the initial call to `Connect()` that was there previously at https://github.com/elastic/beats/blob/367d94e1dc40a82713d651735eac7921bb584474/libbeat/publisher/pipeline/client_worker.go#L143

In the case of Kafka the Connect call never happening makes the output panic on the first call to `Publish` because there is no producer.

~Raising without tests to make sure nothing else was missed while I figure out the best test for this that isn't totally contrived, because the problem is in the publisher pipeline and the output specific tests don't hook in at that level. This is how the problem was originally missed.~

The PR has been updated with an integration test to reproduce the problem: https://github.com/elastic/beats/pull/41824#issuecomment-2506777622

### CODE REVIEW

**Comment by cmacknz** [comment]
No changelog needed because the bug was never released. It is only on main, 8.x, and the 8.17 branch.

**Comment by cmacknz** [comment]
I think the only thing that would catch this is an integration test, I am going to try to write one that looks like https://github.com/elastic/beats/blob/367d94e1dc40a82713d651735eac7921bb584474/libbeat/tests/integration/elasticsearch_test.go#L56

I don't think we need actual Kafka for the regression test, just to observe that we tried to connect to it. I will either bind nothing to 9092 or maybe I'll make or find a small kafka broker.

**Comment by cmacknz** [comment]
https://github.com/elastic/sarama/blob/c7eabfcee7e5bcd7d0071f0ece4d6bec8c33928a/mockbroker.go#L31 might be what we need

**Review by rdner — APPROVED** [review]
We need tests in a follow up. Merging this PR should not close the reported issue.

**Review by mauri870 — APPROVED** [review]

**Comment by cmacknz** [comment]
I have an integration test locally using a Mock Kafka broker that catches the problem, I am just polishing it. Will push soon.

**Comment by cmacknz** [comment]
Updated with an integration test that reproduces the problem using Sarama's MockBroker to ensure we can successfully publish a message from mockbeat using the entire Beat pipeline, without needing to configure a real Kafka instance.

File: libbeat/outputs/kafka/client.go [hunk 1]
```diff
@@ -113,7 +113,7 @@ func newKafkaClient(
 	return c, nil
 }
 
-func (c *client) Connect() error {
+func (c *client) Connect(_ context.Context) error {
```
Comment (cmacknz) [inline libbeat/outputs/kafka/client.go:116]: If you revert this change and run the integration test, it will fail with the panic in the original issue.

File: libbeat/tests/integration/kafka_test.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+	"time"
+
+	"github.com/Shopify/sarama"
+)
+
+var (
+	// https://github.com/elastic/sarama/blob/c7eabfcee7e5bcd7d0071f0ece4d6bec8c33928a/config_test.go#L14-L17
+	// The version of MockBroker used when this test was written only supports the lowest protocol version by default.
+	// Version incompatibilities will result in message decoding errors between the mock and the beat.
+	kafkaVersion = sarama.MinVersion
```
Comment (cmacknz) [inline libbeat/tests/integration/kafka_test.go:34]: Discovering the need for this cost me 3 hours 🙃
